overlord/snapstate/snapstate.go

```diff
@@ -2067,44 +2067,67 @@ func splitRefresh(st *state.State, essential, nonEssential []minimalInstallInfo,
 		}
 	}
 
-	// add dependencies between apps and the boot base, if required
-	var crossSetDependency bool
-	for _, base := range essential {
-		if base.Type() != snap.TypeBase {
-			continue
+	deviceCtx, err := DevicePastSeeding(st, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// we might need to setup dependencies between other apps and the model base,
+	// so find its last refresh task
+	var modelBaseEndTask *state.Task
+	var modelBaseInEssentialTaskset bool
+	if strutil.ListContains(allUpdated, deviceCtx.Base()) {
+		// the base might be in the non-essential tss, if there are no kernel/gadget updates
+		var baseIncTss []*state.TaskSet
+		if len(essentialUpdated) > 0 {
+			baseIncTss = essentialTss.Refresh
+			modelBaseInEssentialTaskset = true
+		} else {
+			baseIncTss = nonEssentialTss.Refresh
+		}
+
+		baseTs, err := maybeFindTasksetForSnap(baseIncTss, deviceCtx.Base())
+		if err != nil {
+			return nil, nil, err
+		}
+		if baseTs == nil {
+			return nil, nil, fmt.Errorf("internal error: cannot find taskset for snap %q as expected", deviceCtx.Base())
 		}
 
+		modelBaseEndTask = baseTs.MaybeEdge(EndEdge)
+		if modelBaseEndTask == nil {
+			return nil, nil, fmt.Errorf("internal error: cannot find last task in base's update taskset")
+		}
+
+		// make note of the last task in case a default provider that depends on
+		// this base is added mid-change
+		modelBaseEndTask.Set("model-base-end", true)
+	}
+
+	// add dependencies between apps and the boot base, if there's a model base in
+	// the essential taskset and an app depends on it
+	var crossSetDependency bool
+	if modelBaseInEssentialTaskset {
 		for _, app := range nonEssential {
-			if app.SnapBase() != base.InstanceName() {
+			// skip if we're not updating this app or its base is not the model base
+			if !strutil.ListContains(nonEssentialUpdated, app.InstanceName()) || app.SnapBase() != deviceCtx.Base() {
 				continue
 			}
 
-			baseTaskset, err := maybeFindTasksetForSnap(essentialTss.Refresh, base.InstanceName())
-			if err != nil {
-				return nil, nil, err
-			}
-
 			appTaskset, err := maybeFindTasksetForSnap(nonEssentialTss.Refresh, app.InstanceName())
 			if err != nil {
 				return nil, nil, err
 			}
-
-			if baseTaskset == nil || appTaskset == nil {
-				// one of the snaps is not being updated
-				continue
+			if appTaskset == nil {
```

Comment:
`maybeFindTasksetForSnap` can return `nil, nil`. But now we error if we don't find it in all cases that it is called (with the exception of [here](https://github.com/snapcore/snapd/blob/master/overlord/snapstate/snapstate.go?plain=1#L2048), but I think that usage is missing a `nil` check? But probably safe since we check for snapd's presence inside of `allUpdated`). Probably should change the function to just return an error if it can't find the task set.